### PR TITLE
Add getdeprecationinfo RPC method to return deprecation block height

### LINKED ISSUE #2828: Add an RPC method for getting the height at which zcashd will auto-senesce
Currently we print the deprecation height on the metrics screen and in the `debug.log`. But users aren't necessarily monitoring either of these (see e.g. #2822).

We should add an RPC method for obtaining the deprecation height of the current `zcashd` version. We can't force users to call it, but wallet GUIs could use it to pre-emptively notify their users.

### PR DESCRIPTION
**Pull request by arcalinea**
Closes #2828 

Returns:
```
{
  "version": xxxxx,                      (numeric) the server version
  "subversion": "/MagicBean:x.y.z[-v]/",     (string) the server subversion string
  "deprecationheight": xxxxx,            (numeric) the deprecation block height
}
```

### CODE REVIEW

File: src/rpcnet.cpp [hunk 1]
```diff
@@ -398,6 +399,34 @@ static UniValue GetNetworksInfo()
...
+UniValue getdeprecationinfo(const UniValue& params, bool fHelp)
+{
+    if (fHelp || params.size() != 0)
+        throw runtime_error(
+            "getdeprecationinfo\n"
+            "Returns an object containing current version and deprecation block height.\n"
+            "\nResult:\n"
+            "{\n"
+            "  \"version\": xxxxx,                      (numeric) the server version\n"
+            "  \"subversion\": \"/MagicBean:x.y.z[-v]/\",     (string) the server subversion string\n"
```
Comment (str4d) [inline src/rpcnet.cpp]: I find the naming of this irritating (it's an extended representation of the version), but it matches existing Bitcoin-inherited RPCs and API designs, so leave it as-is.

File: src/rpcnet.cpp [hunk 2]
```diff
@@ -398,6 +399,34 @@ static UniValue GetNetworksInfo()
...
+            "  \"deprecationheight\": xxxxx,            (numeric) the deprecation block height\n"
```
Comment (str4d) [inline src/rpcnet.cpp]: `the block height at which this version will deprecate and shut down (unless -disabledeprecation is set)`

(modulo line lengths - look at existing RPC help messages for comparison)

File: src/rpcnet.cpp [hunk 3]
```diff
@@ -398,6 +399,34 @@ static UniValue GetNetworksInfo()
     return networks;
 }
 
+UniValue getdeprecationinfo(const UniValue& params, bool fHelp)
```
Comment (str4d) [inline src/rpcnet.cpp]: Hmm... would this make more sense in `rpcmisc.cpp`? If I had to pick a specific category then I *would* put it in `net`, but I think it's closer to upstream's `getmemoryinfo`, in that it is about the local node. I'm not fussed about bikeshedding this though.

File: src/rpcnet.cpp [hunk 4]
```diff
@@ -398,6 +399,34 @@ static UniValue GetNetworksInfo()
...
+            "  \"version\": xxxxx,                      (numeric) the server version\n"
+            "  \"subversion\": \"/MagicBean:x.y.z[-v]/\",     (string) the server subversion string\n"
+            "  \"deprecationheight\": xxxxx,            (numeric) the deprecation block height\n"
+            "}\n"
+            "\nExamples:\n"
+            + HelpExampleCli("getdeprecationinfo", "")
+            + HelpExampleRpc("getdeprecationinfo", "")
+        );
+
+    LOCK(cs_main);
```
Comment (str4d) [inline src/rpcnet.cpp]: No need to take the lock, as this function only depends on hard-coded constants.

File: src/rpcnet.cpp (hunk 3 above)
Comment (arcalinea) [reply src/rpcnet.cpp]: I put it in `rpcnet.cpp` because it also returns version info, duplicating some of what `getnetworkinfo` does

**Comment by arcalinea** [comment]
Addressed @str4d's comments

File: src/rpcnet.cpp [hunk 5]
```diff
@@ -398,6 +399,33 @@ static UniValue GetNetworksInfo()
...
+            "  \"deprecationheight\": xxxxx,            (numeric) the block height at which this version will deprecate and shut down (unless -disabledeprecation is set)\n"
+            "}\n"
+            "\nExamples:\n"
+            + HelpExampleCli("getdeprecationinfo", "")
+            + HelpExampleRpc("getdeprecationinfo", "")
+        );
+
+
+    UniValue obj(UniValue::VOBJ);
+    obj.push_back(Pair("version",       CLIENT_VERSION));
```
Comment (str4d) [inline src/rpcnet.cpp]: Remove indentation, since it doesn't align with anything else.

**Comment by arcalinea** [comment]
Huh, good to know. Ok, will update help message and check for mainnet before showing.

File: src/rpcnet.cpp [hunk 6]
```diff
@@ -398,6 +399,33 @@ static UniValue GetNetworksInfo()
     return networks;
 }
 
+UniValue getdeprecationinfo(const UniValue& params, bool fHelp)
+{
+    const CChainParams& chainparams = Params();
+    if (fHelp || params.size() != 0 || chainparams.NetworkIDString() != "mainnet")
```
Comment (str4d) [inline src/rpcnet.cpp]: `Params().NetworkIDString() != "main"`

**Comment by arcalinea** [comment]
Squashed

**Comment by bitcartel** [comment]
@zkbot r+

**Pull request merged**